File: legged_gym/resources/robots/g1/urdf_to_mjcf.cpp
// 简单的 URDF -> MJCF 转换器（基于 MuJoCo 自带的 URDF 解析能力）
// 用法： urdf_to_mjcf <input.urdf> [output.xml]

#include <mujoco/mujoco.h>
#include <iostream>
#include <string>
#include <cstring>

static std::string deriveOutputPathFromInput(const std::string& inputPath) {
    // 将 .urdf 或 .xml 后缀改为 _converted.xml；否则追加 .xml
    const std::string::size_type dot = inputPath.find_last_of('.') ;
    if (dot != std::string::npos) {
        const std::string ext = inputPath.substr(dot);
        if (ext == ".urdf" || ext == ".URDF" || ext == ".xml" || ext == ".XML") {
            return inputPath.substr(0, dot) + "_converted.xml";
        }
    }
    return inputPath + ".xml";
}

int main(int argc, char** argv) {
    if (argc < 2) {
        std::cerr << "用法: " << argv[0] << " <input.urdf> [output.xml]" << std::endl;
        std::cerr << "说明: 依赖 MuJoCo 对 URDF 的解析（URDF 中可包含 <mujoco><compiler .../> 指令）。" << std::endl;
        return 1;
    }

    const std::string inputPath = argv[1];
    const std::string outputPath = (argc >= 3) ? std::string(argv[2]) : deriveOutputPathFromInput(inputPath);

    char error[1024] = {0};
    // 直接从文件系统加载。若需要自定义 VFS，可在此构造 mjVFS。
    mjModel* model = mj_loadXML(inputPath.c_str(), nullptr, error, sizeof(error));
    if (!model) {
        std::cerr << "加载失败: " << inputPath << std::endl;
        if (std::strlen(error) > 0) {
            std::cerr << "错误信息: " << error << std::endl;
        }
        return 2;
    }

    if (!mj_saveLastXML(outputPath.c_str(), model, error, sizeof(error))) {
        std::cerr << "保存 MJCF 失败: " << outputPath << std::endl;
        if (std::strlen(error) > 0) {
            std::cerr << "错误信息: " << error << std::endl;
        }
        mj_deleteModel(model);
        return 3;
    }

    std::cout << "成功转换为 MJCF: " << outputPath << std::endl;
    mj_deleteModel(model);
    return 0;
}

File: legged_gym/resources/robots/urdf_to_mjcf.py
#!/usr/bin/env python3
"""
URDF -> MJCF XML converter using MuJoCo's built-in URDF loader.

Usage:
  python3 urdf_to_mjcf.py <input.urdf> [-o <output.xml>]

Notes:
- Relative mesh paths are resolved by temporarily chdir to the URDF's directory.
- URDF may include <mujoco><compiler meshdir=.../> directives which MuJoCo honors.
"""

import argparse
import os
import sys
import mujoco  # pip install mujoco


def derive_output_path(input_path: str) -> str:
    base, ext = os.path.splitext(input_path)
    if ext.lower() in (".urdf", ".xml"):
        return base + "_converted.xml"
    return input_path + ".xml"


def main() -> int:
    parser = argparse.ArgumentParser(description="Convert URDF to MJCF XML using MuJoCo")
    parser.add_argument("input", help="Path to input .urdf (or .xml)")
    parser.add_argument("-o", "--output", help="Path to output MJCF .xml")
    args = parser.parse_args()

    input_path = os.path.abspath(args.input)
    if not os.path.isfile(input_path):
        print(f"[Error] Input not found: {input_path}", file=sys.stderr)
        return 1

    output_path = os.path.abspath(args.output) if args.output else os.path.abspath(derive_output_path(input_path))
    os.makedirs(os.path.dirname(output_path), exist_ok=True)

    # Load model with paths resolved relative to URDF directory
    urdf_dir = os.path.dirname(input_path)
    urdf_file = os.path.basename(input_path)
    prev_cwd = os.getcwd()
    try:
        os.chdir(urdf_dir)
        model = mujoco.MjModel.from_xml_path(urdf_file)
    except Exception as ex:
        print(f"[Error] Failed to load URDF: {input_path}\n{ex}", file=sys.stderr)
        return 2
    finally:
        os.chdir(prev_cwd)

    # Save compiled MJCF (MuJoCo 2.3.x Python API: mj_saveLastXML(filename, model) -> None)
    try:
        mujoco.mj_saveLastXML(output_path, model)
    except Exception as ex:
        print(f"[Error] Failed to save MJCF to: {output_path}\n{ex}", file=sys.stderr)
        return 3

    print(f"[OK] Saved MJCF: {output_path}")
    return 0


if __name__ == "__main__":
    sys.exit(main())

